app/scripts/models/ThreeViewer.js

```diff
@@ -0,0 +1,306 @@
+define(['backbone', 'Viewport', 'staticHelpers', 'GeometryExport'], function (Backbone, Viewport, staticHelpers, GeometryExport) {
+
+    var colors = {
+        selected: 0x00FFFF,
+        notSelected: 0x999999,
+        notSelectedLine: 0x000000
+    };
+
+    function colorSelected(threeGeom, selected) {
+
+        var partMat,
+            meshMat,
+            lineMat;
+
+        if (selected) {
+            partMat = new THREE.PointCloudMaterial({ color: colors.selected, size: 5, sizeAttenuation: false, side: THREE.DoubleSide });
+            meshMat = new THREE.MeshPhongMaterial({ color: colors.selected, side: THREE.DoubleSide });
+            lineMat = new THREE.LineBasicMaterial({ color: colors.selected });
+        }
+        else {
+            partMat = new THREE.PointCloudMaterial({ color: colors.notSelected, size: 5, sizeAttenuation: false, side: THREE.DoubleSide });
+            meshMat = new THREE.MeshPhongMaterial({ color: colors.notSelected, side: THREE.DoubleSide });
+            lineMat = new THREE.LineBasicMaterial({ color: colors.notSelectedLine });
+        }
+
+        setMaterials(threeGeom, partMat, meshMat, lineMat);
+    };
+
+    function setMaterials(threeGeom, partMat, meshMat, lineMat) {
+
+        threeGeom.traverse(function (ele) {
+            if (ele instanceof THREE.Mesh) {
+                ele.material = meshMat;
+            } else if (ele instanceof THREE.Line) {
+                ele.material = lineMat;
+            } else {
+                ele.material = partMat;
+            }
+        });
+    };
+
+    function nodeSelected(e) {
+        if (!this.idMap[e.id])
+            return;
+
+        colorSelected(this.idMap[e.id], e.selected);
+    };
+
+    function nodeVisible(e) {
+        if (!this.idMap[e.id])
+            return;
+
+        if (e.visible)
+            this.viewPort.scene.add(this.idMap[e.id]);
+        else
+            this.viewPort.scene.remove(this.idMap[e.id]);
+    };
+
+    function nodeRemove(e) {
+        if (!this.idMap[e.id])
+            return;
+
+        this.viewPort.scene.remove(this.idMap[e.id]);
+        delete this.idMap[e.id];
+    };
+
+    function changeWorkspace(e) {
+        e.ids.forEach(function (node) {
+            if (!this.idMap[node])
+                return;
+
+            if (e.visible)
+                this.viewPort.scene.add(this.idMap[node]);
+            else
+                this.viewPort.scene.remove(this.idMap[node]);
```

Comment:
Hi @pbidenko,

Generally we should use `traverse({visible = false})` if we only want to hide the geometry.  It is far faster than adding/removing from the scene.  We SHOULD remove the geometry when the workspace is removed, though.